src/sync/ms_queue.rs

```diff
@@ -174,8 +174,15 @@ impl<T> MsQueue<T> {
                         unsafe {
                             // signal the thread
                             (*signal).data = Some(cache.into_data());
+                            let thread = (*signal).thread.clone();
+
+                            // It's important that we store the data and get the thread that needs
+                            // to be woken up before setting `ready` to `true`.
+                            // FIXME(stjepang): Use `compiler_fence` when it gets stabilized.
+                            fence(SeqCst);
```

Comment:
On another thought, I think using a fence here is fairly convoluted. I've pushed a new commit that simplifies this code a bit.

Basically, I just want `let thread = (*signal).thread.clone()` to happen before `(*signal).ready.store(true, Release)`.

The `Release` ordering will be enough (if we later on load `ready` using `Acquire`).